spetlr-org/spetlr
# Migrate to unity catalog and terraform

## What type of PR is this?

- Refactor
- Bug Fix
- Optimization
- Deployment (Azure, CI/CD, PowerShell, ...)

## Description

- Introduce a test deployment with full Unity Catalog
- Change the spetlr version to be the same as the Databricks Runtime Version used in testing - 14.3.* for now
- Test only on 1 DBR version
- Make Speltr UC compatible (see below)
- Change the test deployment to be completely in terraform
- SQL server test were not ported to terraform yet, we invite future PRs to re-introduce this part of the tests.
- 

### Functional changes to Speltr

- The DeltaHandle by default no longer reads or writes by path. All read and writes are done using the name. 
         Unity Catalog tables are either managed, residing in UC managed storage or external (i.e. residing in storage that is known to UC, but not exposed to the user.) In both cases UC wants the user to go trough UC when accessing the table so that permissions can be applied. Therefore all direct access to the underlying data (regardless if that may save a few milliseconds) is not allowed.
- SQL server classes and some other code areas have been marked as deprecated for the simple reason that they are untested. The code has not been changed from previous versions and can still be used.
- Azure tag access is just a convenience method to access azure workspace resource tags in code.
- The Version has been increased to 14.3.0 to align with the DBR version that we test with.

### Deployment

Deployment is split into several stages:
- Permanent resources (metastore_and_pemanent) deploys the metastore and the permanent resources (Cosmos account)
- "integration_resources" deploys all azure resources, the captain service principal and the workspace.
- "integration_databricks" then uses that workspace API and deploys the catalog, workspace association, external locations and permissions so that the SPN can deploy the test job
- Test job submission is a simple step. The spetlr-tools repos could in fact use the pipeline to here and deploy its own set of commands at this point.

## Next steps after this

- Sql Server tests were taken out to descope the transformation. The powershell code has been left in the repo to serve as a reference for when these test resources are ported into terraform.
- Create a cleanup pipeline that deletes the captains and catalogs from databricks. Currently the tests leak 1 catalog and 1 databricks service principal on each run. They can be deleted by administrators. A cleanup script is needed but not critical.
